File: Src/static/js/preview_programas_fichas.js
// static/js/preview_programas_fichas.js
$(document).ready(function () {
  // ------------ Helpers ------------
  function initSelect2() {
    $("#edit-oferta, #edit-lider").select2({
      placeholder: "Seleccionar",
      allowClear: true,
      dropdownParent: $("#modal-edit-ficha"),
    });
  }

  // ------------ Inicializaciones ------------
  initSelect2();
  $("#table-programas").DataTable();
  $("#table-fichas").DataTable();
  document
    .querySelectorAll('[data-bs-toggle="tooltip"]')
    .forEach((el) => new bootstrap.Tooltip(el));

  // ------------ Toggle + Filtrar Fichas ------------
  $(".btn-view-fichas").on("click", function () {
    const progId = $(this).data("program-id");
    const $sec = $("#fichas-section");
    const $rows = $("#table-fichas tbody tr");

    if ($sec.is(":visible") && $sec.data("last") == progId) {
      $sec.slideUp().removeData("last");
    } else {
      $sec.slideDown();
      $rows.hide().filter(`[data-program-id="${progId}"]`).show();
      $sec.data("last", progId);
    }
  });

  // ------------ Editar Programa ------------
  $(".btn-edit-program").on("click", function () {
    const row = $(this).closest("tr");
    const programId = $(this).data("program-id");

    $("#form-edit-program").html(`
      <input type="hidden" name="id_prog" value="${programId}">
      <div class="mb-3">
        <label class="form-label">Nombre</label>
        <input type="text" class="form-control" name="nombre" value="${row
          .find("td:eq(2)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Sigla</label>
        <input type="text" class="form-control" name="sigla" value="${row
          .find("td:eq(3)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Trimestres</label>
        <input type="number" class="form-control" name="cant_trimestres" value="${row
          .find("td:eq(4)")
          .text()}">
      </div>
      <button type="submit" class="btn btn-success">Guardar Cambios</button>
    `);
  });

  // ------------ Editar Ficha ------------
  $(".btn-edit-ficha").on("click", function () {
    const row = $(this).closest("tr");
    const fichaId = $(this).data("ficha-id");
    const modCode = row.data("modalidad"); // 0 o 1
    const ofertaCode = row.data("oferta"); // 0 o 1

    $("#form-edit-ficha").html(`
      <input type="hidden" name="numero_ficha" value="${fichaId}">
      <div class="mb-3">
        <label class="form-label">Jornada</label>
        <input type="text" name="jornada" class="form-control" value="${row
          .find("td:eq(3)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Código Sede</label>
        <input type="text" name="codigo_sede" class="form-control" value="${row
          .find("td:eq(4)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Inicio Lectiva</label>
        <input type="date" name="fecha_inicio_lectiva" class="form-control" value="${row
          .find("td:eq(5)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Fin Lectiva</label>
        <input type="date" name="fecha_fin_lectiva" class="form-control" value="${row
          .find("td:eq(6)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Inicio Inducción</label>
        <input type="date" name="fecha_inicio_induccion" class="form-control" value="${row
          .find("td:eq(7)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Fin Inducción</label>
        <input type="date" name="fecha_fin_induccion" class="form-control" value="${row
          .find("td:eq(8)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Instructor (ID)</label>
        <input type="number" name="id_instructor" class="form-control" value="${row
          .find("td:eq(9)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Intensidad (horas)</label>
        <input type="number" name="intensidad" class="form-control" value="${row
          .find("td:eq(10)")
          .text()}">
      </div>
      <div class="mb-3">
        <label class="form-label">Modalidad</label>
        <select id="edit-modalidad" name="modalidad" class="form-select">
          <option value="0">Virtual</option>
          <option value="1">Presencial</option>
        </select>
      </div>
      <div class="mb-3">
        <label class="form-label">Oferta</label>
        <select id="edit-oferta" name="oferta" class="form-select">
          <option value="0">Abierta</option>
          <option value="1">Cerrada</option>
        </select>
      </div>
      <div class="mb-3">
        <label class="form-label">Líder de Ficha</label>
        <select id="edit-lider" name="lider" class="form-select">
          <!-- Opciones dinámicas de instructores -->
        </select>
      </div>
      <button type="submit" class="btn btn-success">Guardar Cambios</button>
    `);

    $("#edit-modalidad").val(modCode);
    $("#edit-oferta").val(ofertaCode);
    initSelect2();
  });

  // ------------ Envío de Formularios ------------
  $("#form-edit-program, #form-edit-ficha").on("submit", function (e) {
    e.preventDefault();
    Swal.fire("Guardado", "Cambios almacenados 👍", "success");
    $("#modal-edit-program, #modal-edit-ficha").modal("hide");
    // TODO: agregar AJAX aquí si lo necesitas
  });
});

File: Src/routes/preview_programas_fichas.py
from flask import Blueprint, render_template, current_app, request, jsonify

preview_programas_fichas_bp = Blueprint(
    'preview_programas_fichas',
    __name__,
    template_folder='../templates'
)

@preview_programas_fichas_bp.route('/preview_programas_fichas')
def preview_programas_fichas():
    mysql = current_app.mysql
    cur = mysql.connection.cursor()

    # 1) Programas
    cur.execute("SELECT id_prog, nombre, sigla, cant_trimestres FROM programas")
    programas = [
        {'id_prog': p[0], 'nombre': p[1], 'sigla': p[2], 'cant_trimestres': p[3]}
        for p in cur.fetchall()
    ]

    # 2) Fichas
    cur.execute("""
        SELECT
            numero_ficha,
            programa,
            jornada,
            codigo_sede,
            DATE_FORMAT(fecha_inicio_lectiva,   '%Y-%m-%d') AS ini_lectiva,
            DATE_FORMAT(fecha_fin_lectiva,       '%Y-%m-%d') AS fin_lectiva,
            DATE_FORMAT(fecha_inicio_induccion, '%Y-%m-%d') AS ini_induccion,
            DATE_FORMAT(fecha_fin_induccion,     '%Y-%m-%d') AS fin_induccion,
            id_instructor,
            intensidad,
            modalidad,
            oferta
        FROM fichas
    """)
    raw = cur.fetchall()
    cur.close()

    fichas = []
    for f in raw:
        # f = tuple: (num, prog, jornada, sede, iniL, finL, iniI, finI, instr, inten, mod, of)
        modalidad_txt = 'Presencial' if f[10] == 1 else 'Virtual'
        oferta_txt    = 'Abierta'    if f[11] == 0 else 'Cerrada'
        fichas.append({
            'numero_ficha':         f[0],
            'programa':             f[1],
            'jornada':              f[2],
            'codigo_sede':          f[3],
            'fecha_inicio_lectiva': f[4],
            'fecha_fin_lectiva':    f[5],
            'fecha_inicio_induccion': f[6],
            'fecha_fin_induccion':    f[7],
            'id_instructor':        f[8],
            'intensidad':           f[9],
            'modalidad':            f[10],
            'oferta':               f[11],
        })

    return render_template(
        'preview_programas_fichas.html',
        programas=programas,
        fichas=fichas
    )

@preview_programas_fichas_bp.route('/preview_programas_fichas/update_programa', methods=['POST'])
def update_programa():
    mysql = current_app.mysql
    cur = mysql.connection.cursor()
    data = request.form
    cur.execute(
        """
        UPDATE programas
        SET nombre = %s,
            sigla  = %s,
            cant_trimestres = %s
        WHERE id_prog = %s
        """,
        (data['nombre'], data['sigla'], data['cant_trimestres'], data['id_prog'])
    )
    mysql.connection.commit()
    cur.close()
    return jsonify(success=True)

@preview_programas_fichas_bp.route('/preview_programas_fichas/update_ficha', methods=['POST'])
def update_ficha():
    mysql = current_app.mysql
    cur = mysql.connection.cursor()
    data = request.form
    # Convertir modalidad/oferta de texto a código si vienes de un select de texto
    mod_code   = 1 if data['modalidad'] == 'Presencial' else 0
    oferta_code= 0 if data['oferta']    == 'Abierta'    else 1

    cur.execute(
        """
        UPDATE fichas
        SET jornada = %s,
            codigo_sede = %s,
            fecha_inicio_lectiva = %s,
            fecha_fin_lectiva = %s,
            fecha_inicio_induccion = %s,
            fecha_fin_induccion = %s,
            id_instructor = %s,
            intensidad = %s,
            modalidad = %s,
            oferta = %s
        WHERE numero_ficha = %s
        """,
        (
            data['jornada'],
            data['codigo_sede'],
            data['fecha_inicio_lectiva'],
            data['fecha_fin_lectiva'],
            data['fecha_inicio_induccion'],
            data['fecha_fin_induccion'],
            data['id_instructor'],
            data['intensidad'],
            mod_code,
            oferta_code,
            data['numero_ficha']
        )
    )
    mysql.connection.commit()
    cur.close()
    return jsonify(success=True)